lib/view_component/compiler.rb

```diff
@@ -103,266 +50,273 @@ def renders_template_for?(variant, format)
+        errors << "More than one #{this_format.upcase} template found#{variant_string} for #{@component}. "
+      end
 
-            errors << "More than one #{this_format.upcase} template found#{variant_string} for #{component_class}. "
-          end
+      if @templates.any? { _1.variant.nil? && !_1.inline_call? } &&
+          @templates.any? { _1.variant.nil? && _1.inline_call? && _1.defined_on_self? }
 
-          if templates.find { |template| template[:variant].nil? } && inline_calls_defined_on_self.include?(:call)
-            errors <<
-              "Template file and inline render method found for #{component_class}. " \
-              "There can only be a template file or inline render method per component."
-          end
+        errors <<
+          "Template file and inline render method found for #{@component}. " \
+          "There can only be a template file or inline render method per component."
+      end
 
-          duplicate_template_file_and_inline_variant_calls =
-            templates.pluck(:variant) & variants_from_inline_calls(inline_calls_defined_on_self)
+      duplicate_template_file_and_inline_call_variants =
+        @templates.select { !_1.inline_call? }.map(&:variant) &
+        @templates.select { _1.inline_call? && _1.defined_on_self? }.map(&:variant)
 
-          unless duplicate_template_file_and_inline_variant_calls.empty?
-            count = duplicate_template_file_and_inline_variant_calls.count
+      unless duplicate_template_file_and_inline_call_variants.empty?
+        count = duplicate_template_file_and_inline_call_variants.count
 
-            errors <<
-              "Template #{"file".pluralize(count)} and inline render #{"method".pluralize(count)} " \
-              "found for #{"variant".pluralize(count)} " \
-              "#{duplicate_template_file_and_inline_variant_calls.map { |v| "'#{v}'" }.to_sentence} " \
-              "in #{component_class}. " \
-              "There can only be a template file or inline render method per variant."
-          end
+        errors <<
+          "Template #{"file".pluralize(count)} and inline render #{"method".pluralize(count)} " \
+          "found for #{"variant".pluralize(count)} " \
+          "#{duplicate_template_file_and_inline_call_variants.map { |v| "'#{v}'" }.to_sentence} " \
+          "in #{@component}. There can only be a template file or inline render method per variant."
+      end
 
-          uniq_variants = variants.compact.uniq
-          normalized_variants = uniq_variants.map { |variant| normalized_variant_name(variant) }
+      variant_pairs =
+        @templates.select { _1.variant.present? }.map { [_1.variant, _1.normalized_variant_name] }.uniq { _1.first }
 
-          colliding_variants = uniq_variants.select do |variant|
-            normalized_variants.count(normalized_variant_name(variant)) > 1
-          end
+      colliding_normalized_variants =
```

Comment:
I'm finding this pretty difficult to follow, and I think it's doing a good number of unnecessary iterations over data that could potentially be avoided with a more imperative approach (I know it's likely this is just complexity we're surfacing as part of the refactor, but feels worth fixing?).


I don't have a significant amount of time to hack on this at the moment, but I wonder if this captures the same behavior (combined with my above suggestion)?

```ruby
        # Keep track of the variants that are colliding due to multiple templates referencing the same normalized variant
        colliding_normalized_variants = Set.new
        variant_pairs = @templates.each_with_object({}) do |template, pairs|
          next unless template.variant
          if pairs[template.normalized_variant_name]
            colliding_normalized_variants << template.normalized_variant_name
          else
            pairs[template.normalized_variant_name] ||= []
          end
          
          pairs[template.normalized_variant_name] << template.normalized_variant_name
        end
```

Then we could use `colliding_normalized_variants` to get the relevant values from the map, I think?

Reply:
Like `variant_pairs.fetch_values(colliding_normalized_variants)`? (which I think could be further optimized to avoid this call too)

Reply:
See https://github.com/ViewComponent/view_component/pull/2091/commits/a2d2f848dee13d207a549207c5f29369b69b69c1, I think I found a nice middle ground here!